nb2kg/handlers.py

```diff
@@ -86,8 +89,23 @@ def get(self, kernel_id, *args, **kwargs):
         self.kernel_id = cast_unicode(kernel_id, 'ascii')
         yield gen.maybe_future(super(WebSocketChannelsHandler, self).get(kernel_id=kernel_id, *args, **kwargs))
 
+    def send_ping(self):
+        if self.stream.closed() and self.ping_callback is not None:
```

Comment:
I'm seeing the following every 30 seconds.  Since the callback is never stopped, this error gets logged every 30 seconds even after the kernel is shutdown...
```
[E 08:11:10.512 NotebookApp] Exception in callback <bound method WebSocketChannelsHandler.send_ping of <nb2kg.handlers.WebSocketChannelsHandler object at 0x1038c29b0>>
    Traceback (most recent call last):
      File "/opt/anaconda3/envs/nb2kg-dev/lib/python3.7/site-packages/tornado/ioloop.py", line 907, in _run
        return self.callback()
      File "/opt/anaconda3/envs/nb2kg-dev/lib/python3.7/site-packages/nb2kg/handlers.py", line 93, in send_ping
        if self.stream.closed() and self.ping_callback is not None:
    AttributeError: 'NoneType' object has no attribute 'closed'
```
It may be trickier than just testing `self.stream` for `None` since the connection code uses 'detach' and I'm not certain how that actually works.

Reply:
This may be an issue with the move to Tornado 6.  See this update in notebook: https://github.com/jupyter/notebook/commit/5828300401f9f34bd9db70a7734ca798e41c1a9b#diff-808f61143410aa59f6ac185458495d66

However, I suspect there's a bug there in that `self.ws_connection` is really not accessible to the `WebSocketMixin` class or I don't understand how multiple inheritance works in Python.  This member variable is defined in `WebSocketHandler`, which is the sibling super-class to `ZMQStreamHandler`.

That said, since NB2KG doesn't use a mixin here, you're probably fine using `self.ws_connection` in the condition.